### PR TITLE
Fix OwnReceive workaround for large request bodies

### PR DESCRIPTION
**Pull request by JulianSprung**
## Purpose

TLDR: processing large request (e.g. 600kB) bodies in an endpoint hangs forever when using the OPAMiddleware.

In https://github.com/busykoala/fastapi-opa/pull/47 a workaround was implemented that handles the issue that arises from trying access a request body multiple times.
However, the current workaround implementation breaks for larger sizes of the request body (±a few hundred kB in my setup), i.e. it remains in an endless loop trying to `await request.body()` with 100% CPU load.

The reason is that in for larger request bodies, the ASGI framework chunks the request body and sets `more_body` flag to `True`, see https://asgi.readthedocs.io/en/latest/specs/www.html#request-receive-event.

In the current implementation this is not handled, so accessing the request body the second time hangs forever.

## Approach

The workaround, i.e. the `OwnReceive` class was modified to check if the request body is chunked, i.e. contains more_body.
If yes, then the additional chunks are added to the buffer.

I also added an (optional) parameter to set a maximum buffer size.

## Checklist for PRs

- [x] There is a Changelog (`/CHANGELOG.md`)
- [x] Version was adapted if necessary (`/pyproject.toml`)
- [x] I tested the feature if necessary (unittests, manual testing)
- [x] If libraries aren't used for all package usages they are extras
- [x] I documented the changes

### CODE REVIEW

**Comment by busykoala** [comment]
Hi @JulianSprung, thanks a lot for the contribution. I'll try to understand the issue and suggested PR as soon as possible. Would you be able to add a test that runs into the issue with the old code and shows that the new code fixes the issue? It would speed up the process of me understanding the issue and prevents future changes that would break it again.

**Comment by busykoala** [comment]
Sorry for my late reply. I wanted to look into the implementation before going on with it closely.

Regarding the tests, I would suggest something like this to avoid having to add uvicorn and running it in a separate thread:

A new endpoint and fixture in conftest.py:
```py
@pytest.fixture
def client():
    opa_host = "http://localhost:8181"
    oidc_auth = AuthenticationDummy()
    opa_config = OPAConfig(authentication=oidc_auth, opa_host=opa_host)

    # existing code ...

    @app.post("/items")
    async def create_item(request: Request):
        data = await request.json()
        return {"msg": f"Received {len(str(data))} bytes"}

    yield TestClient(app)

@pytest.fixture
def large_body() -> Callable[[], AsyncGenerator[Dict[str, Any], None]]:
    """Fixture to generate a large request body in chunks."""
    async def generate():
        body = b'{"input": {"item_id": 1, "data": "' + b"a" * 5 * 1024 * 1024 + b'"}}'
        for i in range(0, len(body), 1024):
            yield {
                "type": "http.request",
                "body": body[i:i+1024],
                "more_body": i + 1024 < len(body),
            }
        yield {"type": "http.request", "body": b"", "more_body": False}
    return generate
```

And then a single test like this:

```py
import pytest
from unittest.mock import patch
from fastapi.testclient import TestClient
import asyncio
from typing import Dict, Any, Callable, AsyncGenerator


async def simulate_request(client: TestClient, large_body):
    """
    Simulate a request to the FastAPI app with a chunked body.

    This function mimics the ASGI protocol behavior for handling large HTTP
    request bodies. The body is split into chunks, and each chunk is yielded
    through the `mock_receive` function. The ASGI app processes these chunks
    until the entire body is consumed.
    """

    async def mock_receive():
        """
        Mimic the ASGI `receive` function to send chunks of the request body.
        """
        async for chunk in large_body():
            # Yield control to the event loop to allow timeout enforcement
            await asyncio.sleep(0)
            yield chunk

    async def mock_send(message):
        """
        Mimic the ASGI `send` function to capture the app's response.
        """
        nonlocal response_body
        if message["type"] == "http.response.body":
            response_body += message.get("body", b"")

    # Define the ASGI scope for the request
    scope = {
        "type": "http",  # HTTP request
        "method": "POST",  # HTTP method
        "path": "/items",  # Target endpoint
        "headers": [(b"content-type", b"application/json")],  # Request headers
        "query_string": b"",  # No query parameters
    }

    response_body = b""  # Buffer to store the response body

    # Pass the ASGI scope, mock_receive, and mock_send to the ASGI app
    await client.app(scope, mock_receive().__anext__, mock_send)
    return response_body.decode("utf-8")


@pytest.mark.asyncio
@pytest.mark.timeout(2)  # Enforce a 2-second timeout on the test
async def test_large_request_body(
    client: TestClient,
    large_body: Callable[[], AsyncGenerator[Dict[str, Any], None]]
) -> None:
    """
    Test that the FastAPI app handles a large request body correctly.

    This test verifies that the app can process a request with a large JSON
    payload, delivered in chunks, and return the expected response. It also
    mocks the OPA middleware to ensure the app processes the request as if it
    were allowed by OPA.
    """
    with patch("fastapi_opa.opa.opa_middleware.requests.post") as req:
        # Mock the OPA middleware to allow the request
        req.return_value.status_code = 200
        req.return_value.json = lambda: {"result": {"allow": True}}

        # Simulate the ASGI request and capture the response
        response = await simulate_request(client, large_body)

        # Verify that the app processes the large body correctly
        assert response == '{"msg":"Received 5242917 bytes"}'
```

This mimics the actual issue without having to rely on a specific asgi server like uvicorn.

Additionally, I would suggest setting a sane default for `max_buffer_size` instead of `None` to avoid that it is not set accidentally and can be used for targeting the app.

One last thing, the pipeline will be running the following steps (speeds up me being able to merge if you already check that in your PR). I'll add it later on in a Makefile to improve the development experience, but it's not yet there...

```bash
poetry run ruff check
poetry run ruff format --check
poetry run vale README.md CONTRIBUTING.md
poetry run pytest
poetry run bandit -r fastapi_opa --exclude="fastapi_opa/example_oidc.py","fastapi_opa/example_saml.py"
```

**Comment by busykoala** [comment]
I've implemented the change in #77 and mentioned you and the PR in there due to no reply anymore here. @JulianSprung I hope this is ok for you. Thanks a lot for the help with this issue!

**Comment by JulianSprung** [comment]
Hey @busykoala ,
Sorry I missed this conversation over the holidays. Thanks for adding the changes in PR77!
